cirq/protocols/json_test.py

```diff
@@ -0,0 +1,446 @@
+
+    # Circuit optimizers are function-like. Only attributes
+    # are ignore_failures, tolerance, and other feature flags
+    'ConvertToCzAndSingleGates',
+    'ConvertToIonGates',
+    'ConvertToNeutralAtomGates',
+    'DropEmptyMoments',
+    'DropNegligible',
+    'EjectPhasedPaulis',
+    'EjectZ',
+    'ExpandComposite',
+    'MergeInteractions',
+    'MergeSingleQubitGates',
+    'PointOptimizer',
+
+    # global objects
+    'CONTROL_TAG',
+    'PAULI_BASIS',
+
+    # abstract, but not inspect.isabstract():
+    'InterchangeableQubitsGate',
+    'Pauli',
+    'SingleQubitGate',
+    'ThreeQubitGate',
+    'TwoQubitGate',
+
+    # protocols:
+    'SupportsApplyChannel',
+    'SupportsApproximateEquality',
+    'SupportsChannel',
+    'SupportsCircuitDiagramInfo',
+    'SupportsConsistentApplyUnitary',
+    'SupportsDecompose',
+    'SupportsDecomposeWithQubits',
+    'SupportsExplicitNumQubits',
+    'SupportsExplicitQidShape',
+    'SupportsMixture',
+    'SupportsParameterization',
+    'SupportsPhase',
+    'SupportsQasm',
+    'SupportsQasmWithArgs',
+    'SupportsQasmWithArgsAndQubits',
+    'SupportsTraceDistanceBound',
+    'SupportsUnitary',
+
+    # mypy types:
+    'OP_TREE',
+    'ParamResolverOrSimilarType',
+    'PauliSumLike',
+    'QubitOrderOrList',
+    'Sweepable',
+    'TParamVal',
+    'ParamDictType',
+
+    # utility:
+    'Unique',
+]
+
+
+def _get_all_public_classes():
+    for cls_name, cls_cls in inspect.getmembers(cirq):
+        if inspect.isfunction(cls_cls) or inspect.ismodule(cls_cls):
+            continue
+
+        if cls_name in SHOULDNT_BE_SERIALIZED:
+            continue
+
+        if not inspect.isclass(cls_cls):
+            # singletons, for instance
+            cls_cls = cls_cls.__class__
+
+        if cls_name.startswith('_'):
+            continue
+
+        if (inspect.isclass(cls_cls) and
+            (inspect.isabstract(cls_cls) or issubclass(cls_cls, abc.ABCMeta))):
+            continue
+
+        yield cls_name, cls_cls
+
+    # extra
+    yield 'complex', complex
+    yield 'ndarray', np.ndarray
+    yield 'Symbol', sympy.Symbol
+
+    # test coverage for `default` paths
+    yield 'dict', dict
+
+
+def test_shouldnt_be_serialized_no_superfluous():
+    # everything in the list should be ignored for a reason
+    names = [
+        name for name, _ in inspect.getmembers(
+            cirq, lambda x: not (inspect.ismodule(x) or inspect.isfunction(x)))
+    ]
+    for name in SHOULDNT_BE_SERIALIZED:
+        assert name in names
+
+
+def test_not_yet_serializable_no_superfluous():
+    # everything in the list should be ignored for a reason
+    names = [
+        name for name, _ in inspect.getmembers(
+            cirq, lambda x: not (inspect.ismodule(x) or inspect.isfunction(x)))
+    ]
+    for name in NOT_YET_SERIALIZABLE:
+        assert name in names
+
+
+def test_mutually_exclusive_blacklist():
+    assert len(set(SHOULDNT_BE_SERIALIZED) & set(NOT_YET_SERIALIZABLE)) == 0
```

Comment:
Make this error message significantly better. Anyone who adds a gate into cirq is going to see this exception. It should explain clearly what they need to do in order to fix it.

Reply:
the particular line you commented on developers will never see unless they are doing explicit json work. 

I've added a very cordial error message to `test_all_roundtrip`, which is where the error will be raised.

Reply:
Nice fixed fix